File: caia-reviewers/src/main/java/com/solidos/caia/reviewers/domain/entities/Invitation.java
package com.solidos.caia.reviewers.domain.entities;

import java.time.LocalDateTime;

import com.solidos.caia.reviewers.domain.valueobjects.AuditMetadata;
import com.solidos.caia.reviewers.domain.valueobjects.InvitationId;
import com.solidos.caia.reviewers.domain.valueobjects.InvitationState;
import com.solidos.caia.reviewers.utils.TokenGenerator;

import lombok.Builder;
import lombok.Getter;

@Builder
@Getter
public class Invitation {
  private InvitationId id;

  private String token;

  private InvitationState state;

  private String message;

  private LocalDateTime respondedAt;

  private AuditMetadata auditMetadata;

  private Conference conference;

  public void changeState(InvitationState state) {
    if (this.state == InvitationState.PENDING && state == InvitationState.PENDING) {
      throw new IllegalStateException("Cannot change state to PENDING");
    }

    if (this.state == InvitationState.PENDING && state != InvitationState.PENDING) {
      this.respondedAt = LocalDateTime.now();
      this.token = null;
    }

    this.state = state;
  }

  public void setToken() {
    this.token = TokenGenerator.generate();
  }

  public void prepareToSend() {
    this.setToken();
    this.changeState(InvitationState.PENDING);
  }
}

File: caia-frontend/src/core/reviewer/data/models/invitation.ts
export interface Invitation {
  id: ID;
  token: string;
  state: InvitationState;
  message: string;
  respondedAt: null;
  auditMetadata: AuditMetadata;
  conference: Conference;
}

export interface AuditMetadata {
  createdAt: Date;
  updatedAt: null;
  deletedAt: null;
}

export interface Conference {
  id: number;
  name: string;
  slug: string;
}

export interface ID {
  userId: number;
  conferenceId: number;
}

export type InvitationState = "ACCEPTED" | "REJECTED" | "PENDING";
